File: apps/herbicide_workbench/workbench/plots.py
#!/usr/bin/env python3
"""
Script: plots.py
Objective: Wrap Python CLTF diagnostic plots for the Streamlit workbench.
Author: Yi Yu
Created: 2026-06-24
Last updated: 2026-06-25
Inputs: CLTF forcing, prediction, simulation, profile, and bulk-density tables.
Outputs: Matplotlib Figure objects for app rendering.
Usage: Import plotting helpers from workbench.plots.
Dependencies: matplotlib, pandas, cltf, workbench
"""

from __future__ import annotations

from matplotlib import font_manager
from matplotlib.axes import Axes
from matplotlib.figure import Figure
from matplotlib.lines import Line2D
from matplotlib.colors import LinearSegmentedColormap
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd

from workbench.config import ensure_cltf_path

ensure_cltf_path()
import cltf.plotting as cltf_plotting  # noqa: E402


ASSESSMENT_LABEL = "Residue assessment"
ASSESSMENT_COLOR = "#7A0177"
_COLORS = ("#0072B2", "#D55E00", "#009E73", "#CC79A7")
CONCENTRATION_CMAP = LinearSegmentedColormap.from_list(
    "cltf_green_yellow_red",
    ["#006837", "#FFFFBF", "#A50026"],
)
_ARIAL_REGISTERED = False


def _preferred_font_family() -> str:
    return "Arial"


def _register_arial_font() -> None:
    global _ARIAL_REGISTERED

    if _ARIAL_REGISTERED:
        return
    installed = {font.name for font in font_manager.fontManager.ttflist}
    if "Arial" not in installed:
        for font_path in font_manager.findSystemFonts(fontext="ttf"):
            if font_path.lower().endswith("arial.ttf"):
                font_manager.fontManager.addfont(font_path)
                break
    _ARIAL_REGISTERED = True


def configure_matplotlib() -> None:
    """Apply app-level matplotlib defaults."""

    _register_arial_font()
    plt.rcParams["font.family"] = _preferred_font_family()
    plt.rcParams["figure.dpi"] = 130
    plt.rcParams["axes.spines.top"] = False
    plt.rcParams["axes.spines.right"] = False


def _require_columns(
    data: pd.DataFrame,
    columns: tuple[str, ...],
    object_name: str,
) -> None:
    missing = sorted(set(columns).difference(data.columns))
    if missing:
        raise ValueError(
            f"{object_name} is missing columns: {', '.join(missing)}"
        )


def _layer_label(depth_top_mm: object, depth_bottom_mm: object) -> str:
    return f"{float(depth_top_mm):g}\N{EN DASH}{float(depth_bottom_mm):g} mm"


def _forcing_x_values(
    forcing: pd.DataFrame,
) -> tuple[pd.Series, str, bool]:
    if "date" in forcing.columns:
        return pd.to_datetime(forcing["date"]), "Date", True
    if "time_days" in forcing.columns:
        return forcing["time_days"], "Elapsed time (days)", False
    raise ValueError("forcing must contain date or time_days")


def add_assessment_line(
    axis: Axes,
    assessment_value: object,
    x_is_date: bool = False,
) -> None:
    """Add the approved vertical residue-assessment marker to one axis."""

    x_value = (
        pd.Timestamp(assessment_value)
        if x_is_date
        else assessment_value
    )
    axis.axvline(
        x_value,
        color=ASSESSMENT_COLOR,
        linestyle="--",
        linewidth=1.8,
        label=ASSESSMENT_LABEL,
    )


def _add_assessment_to_figure(
    figure: Figure,
    assessment_value: object | None,
    x_is_date: bool = False,
) -> Figure:
    if assessment_value is None:
        return figure
    for axis in figure.axes:
        add_assessment_line(axis, assessment_value, x_is_date=x_is_date)
    return figure


def plot_climate_forcing(
    forcing: pd.DataFrame,
    assessment_day: int | float | None = None,
    assessment_date: object | None = None,
) -> Figure:
    """Plot climate forcing and mark the residue assessment date/day."""

    configure_matplotlib()
    figure = cltf_plotting.plot_climate_forcing(forcing)
    if assessment_date is not None and "date" in forcing.columns:
        return _add_assessment_to_figure(
            figure,
            assessment_date,
            x_is_date=True,
        )
    return _add_assessment_to_figure(figure, assessment_day)


def plot_water_forcing(
    forcing: pd.DataFrame,
    assessment_day: int | float | None = None,
    assessment_date: object | None = None,
) -> Figure:
    """Plot daily rain, PET, and infiltration on one forcing axis."""

    configure_matplotlib()
    _require_columns(
        forcing,
        ("rain_mm", "pet_mm", "daily_infiltration_mm"),
        "forcing",
    )
    x_values, x_label, x_is_date = _forcing_x_values(forcing)
    figure, axis = plt.subplots(figsize=(8.5, 3.6), constrained_layout=True)
    axis.bar(
        x_values,
        forcing["rain_mm"],
        color="#0072B2",
        alpha=0.60,
        label="Rain",
    )
    axis.plot(
        x_values,
        forcing["pet_mm"],
        color="#D55E00",
        linewidth=1.8,
        label="PET",
    )
    axis.plot(
        x_values,
        forcing["daily_infiltration_mm"],
        color="#009E73",
        linewidth=1.8,
        label="Daily infiltration",
    )
    axis.set_xlabel(x_label)
    axis.set_ylabel("Daily water (mm)")
    axis.legend(frameon=False, ncols=3, fontsize=9)
    if assessment_date is not None and x_is_date:
        add_assessment_line(axis, assessment_date, x_is_date=True)
    elif assessment_day is not None:
        add_assessment_line(axis, assessment_day)
    return figure


def plot_cumulative_infiltration(
    forcing: pd.DataFrame,
    assessment_day: int | float | None = None,
    assessment_date: object | None = None,
) -> Figure:
    """Plot cumulative infiltration against time."""

    configure_matplotlib()
    _require_columns(forcing, ("cumulative_infiltration_mm",), "forcing")
    x_values, x_label, x_is_date = _forcing_x_values(forcing)
    values = forcing["cumulative_infiltration_mm"].to_numpy(dtype=float)
    if not np.any(np.isfinite(values)):
        raise ValueError("forcing contains no finite cumulative infiltration")

    figure, axis = plt.subplots(figsize=(8.5, 3.6), constrained_layout=True)
    axis.plot(
        x_values,
        values,
        color="#009E73",
        linewidth=2,
    )
    axis.set_xlabel(x_label)
    axis.set_ylabel("Cumulative infiltration (mm)")
    if assessment_date is not None and x_is_date:
        add_assessment_line(axis, assessment_date, x_is_date=True)
    elif assessment_day is not None:
        add_assessment_line(axis, assessment_day)
    return figure


def plot_cumulative_infiltration_histogram(
    forcing: pd.DataFrame,
) -> Figure:
    """Compatibility wrapper for the time-series cumulative infiltration plot."""

    return plot_cumulative_infiltration(forcing)


def plot_observed_fitted(
    predictions: pd.DataFrame,
    assessment_day: int | float | None = None,
) -> Figure:
    """Plot replicate observations, fitted values, and an assessment marker."""

    configure_matplotlib()
    figure = cltf_plotting.plot_observed_fitted(predictions)
    return _add_assessment_to_figure(figure, assessment_day)


def _violin_values(values: np.ndarray) -> np.ndarray:
    values = values[np.isfinite(values) & (values > 0)]
    if values.size == 0:
        return values
    if values.size >= 2 and np.nanmax(values) > np.nanmin(values):
        return values
    center = float(values[0])
    spread = max(center * 0.015, 1e-9)
    return np.array([center - spread, center, center + spread])


def plot_observation_violin(
    predictions: pd.DataFrame,
    simulation: pd.DataFrame | None = None,
    assessment_day: int | float | None = None,
) -> Figure:
    """Plot observation violins with jittered replicates and dashed fitted lines."""

    configure_matplotlib()
    _require_columns(
        predictions,
        (
            "days_since_application",
            "depth_top_mm",
            "depth_bottom_mm",
            "analysis_concentration_ug_kg",
            "predicted_concentration_ug_kg",
        ),
        "predictions",
    )
    positive = np.concatenate(
        (
            predictions["analysis_concentration_ug_kg"].to_numpy(dtype=float),
            predictions["predicted_concentration_ug_kg"].to_numpy(dtype=float),
        )
    )
    if not np.any(np.isfinite(positive) & (positive > 0)):
        raise ValueError("predictions contain no positive concentrations")

    data = predictions.copy()
    data["layer"] = [
        _layer_label(top, bottom)
        for top, bottom in zip(data["depth_top_mm"], data["depth_bottom_mm"])
    ]
    layer_order = list(dict.fromkeys(data["layer"].astype(str)))
    days = np.sort(data["days_since_application"].unique().astype(float))
    if days.size > 1:
        day_step = float(np.min(np.diff(days)))
        offset_width = max(day_step * 0.08, 0.75)
    else:
        offset_width = 0.75
    layer_offsets = {
        layer: (index - (len(layer_order) - 1) / 2.0) * offset_width
        for index, layer in enumerate(layer_order)
    }
    colors = {
        layer: _COLORS[index % len(_COLORS)]
        for index, layer in enumerate(layer_order)
    }

    figure, axis = plt.subplots(figsize=(8.5, 4.6), constrained_layout=True)
    rng = np.random.default_rng(20260624)
    for layer in layer_order:
        layer_data = data.loc[data["layer"].eq(layer)]
        for day, day_data in layer_data.groupby("days_since_application"):
            observed = day_data["analysis_concentration_ug_kg"].to_numpy(dtype=float)
            violin_values = _violin_values(observed)
            if violin_values.size == 0:
                continue
            position = float(day) + layer_offsets[layer]
            parts = axis.violinplot(
                [violin_values],
                positions=[position],
                widths=max(offset_width * 0.75, 0.25),
                showmeans=False,
                showmedians=False,
                showextrema=False,
            )
            for body in parts["bodies"]:
                body.set_facecolor(colors[layer])
                body.set_edgecolor(colors[layer])
                body.set_alpha(0.22)

            observed = observed[np.isfinite(observed) & (observed > 0)]
            jitter = rng.normal(0.0, max(offset_width * 0.07, 0.03), observed.size)
            axis.scatter(
                np.full(observed.size, position) + jitter,
                observed,
                color=colors[layer],
                edgecolors="#ffffff",
                linewidths=0.4,
                s=28,
                alpha=0.88,
                zorder=3,
            )

        fitted = (
            layer_data.groupby("days_since_application", as_index=False)[
                "predicted_concentration_ug_kg"
            ]
            .mean()
            .sort_values("days_since_application")
        )
        fitted = fitted.loc[
            np.isfinite(fitted["predicted_concentration_ug_kg"])
            & fitted["predicted_concentration_ug_kg"].gt(0)
        ]
        axis.plot(
            fitted["days_since_application"],
            fitted["predicted_concentration_ug_kg"],
            color=colors[layer],
            linestyle="--",
            linewidth=2,
            label=f"{layer} fitted",
        )

    if simulation is not None:
        _require_columns(
            simulation,
            (
                "time_days",
                "concentration_top_ug_kg",
                "concentration_bottom_ug_kg",
            ),
            "simulation",
        )
        ordered_simulation = simulation.sort_values("time_days")
        simulation_columns = [
            ("Top simulation", "concentration_top_ug_kg", _COLORS[0]),
            ("Bottom simulation", "concentration_bottom_ug_kg", _COLORS[1]),
        ]
        for label, column, color in simulation_columns:
            values = ordered_simulation[column].to_numpy(dtype=float)
            keep = np.isfinite(values) & (values > 0)
            if np.any(keep):
                axis.plot(
                    ordered_simulation["time_days"].to_numpy(dtype=float)[keep],
                    values[keep],
                    color=color,
                    linestyle="-",
                    linewidth=1.7,
                    alpha=0.72,
                    label=label,
                )

    axis.set_yscale("log")
    axis.set_xlabel("Days since application")
    axis.set_ylabel("Resident concentration (\N{MICRO SIGN}g/kg)")
    axis.legend(
        handles=[
            Line2D(
                [0],
                [0],
                color=colors[layer],
                linestyle="--",
                linewidth=2,
                label=f"{layer} fitted",
            )
            for layer in layer_order
        ] + [
            Line2D(
                [0],
                [0],
                color=_COLORS[0],
                linestyle="-",
                linewidth=1.7,
                label="Top simulation",
            ),
            Line2D(
                [0],
                [0],
                color=_COLORS[1],
                linestyle="-",
                linewidth=1.7,
                label="Bottom simulation",
            ),
        ],
        frameon=False,
        fontsize=9,
    )
    if assessment_day is not None:
        add_assessment_line(axis, assessment_day)
    return figure


def _time_edges(time_values: np.ndarray) -> np.ndarray:
    if time_values.size == 1:
        return np.array([time_values[0] - 0.5, time_values[0] + 0.5])
    midpoints = (time_values[:-1] + time_values[1:]) / 2.0
    first = time_values[0] - (time_values[1] - time_values[0]) / 2.0
    last = time_values[-1] + (time_values[-1] - time_values[-2]) / 2.0
    return np.concatenate(([first], midpoints, [last]))


def _depth_edges(depth_values: np.ndarray) -> np.ndarray:
    if depth_values.size == 1:
        return np.array([max(0.0, depth_values[0] - 0.5), depth_values[0] + 0.5])
    midpoints = (depth_values[:-1] + depth_values[1:]) / 2.0
    first = max(0.0, depth_values[0] - (depth_values[1] - depth_values[0]) / 2.0)
    last = depth_values[-1] + (depth_values[-1] - depth_values[-2]) / 2.0
    return np.concatenate(([first], midpoints, [last]))


def _numeric_series(values: pd.Series | pd.Index) -> np.ndarray:
    return pd.to_numeric(values, errors="coerce").to_numpy(
        dtype=float,
        na_value=np.nan,
    )


def _numeric_matrix(values: pd.DataFrame | np.ndarray) -> np.ndarray:
    if isinstance(values, pd.DataFrame):
        numeric = values.apply(pd.to_numeric, errors="coerce")
        result = numeric.to_numpy(dtype=float, na_value=np.nan)
    else:
        result = np.asarray(values, dtype=float)
    return np.where(np.isfinite(result), result, np.nan)


def plot_simulation_heatmap(
    simulation: pd.DataFrame,
    top_depth_mm: float,
    bottom_depth_mm: float,
    assessment_day: int | float | None = None,
) -> Figure:
    """Plot simulated resident concentration as a depth heatmap."""

    configure_matplotlib()
    if {"time_days", "depth_mm", "concentration_ug_kg"} <= set(simulation.columns):
        ordered = simulation.sort_values(["depth_mm", "time_days"])
        pivot = ordered.pivot_table(
            index="depth_mm",
            columns="time_days",
            values="concentration_ug_kg",
            aggfunc="mean",
        ).sort_index().sort_index(axis=1)
        time_values = _numeric_series(pivot.columns)
        depth_values = _numeric_series(pivot.index)
        concentration = _numeric_matrix(pivot)
        x_edges = _time_edges(time_values)
        y_edges = _depth_edges(depth_values)
    else:
        _require_columns(
            simulation,
            (
                "time_days",
                "concentration_top_ug_kg",
                "concentration_bottom_ug_kg",
            ),
            "simulation",
        )
        ordered = simulation.sort_values("time_days")
        time_values = _numeric_series(ordered["time_days"])
        concentration = np.vstack(
            [
                _numeric_series(ordered["concentration_top_ug_kg"]),
                _numeric_series(ordered["concentration_bottom_ug_kg"]),
            ]
        )
        x_edges = _time_edges(time_values)
        y_edges = np.array([0.0, float(top_depth_mm), float(bottom_depth_mm)])
    concentration = _numeric_matrix(concentration)

    figure, axis = plt.subplots(figsize=(8.5, 4.5), constrained_layout=True)
    mesh = axis.pcolormesh(
        x_edges,
        y_edges,
        concentration,
        cmap=CONCENTRATION_CMAP,
        shading="flat",
    )
    axis.set_ylim(float(bottom_depth_mm), 0.0)
    axis.set_xlabel("Days since application")
    axis.set_ylabel("Soil depth (mm)")
    axis.axhline(float(top_depth_mm), color="#404040", linestyle=":", linewidth=1)
    if assessment_day is not None:
        add_assessment_line(axis, assessment_day)
    colorbar = figure.colorbar(
        mesh,
        ax=axis,
        orientation="horizontal",
        pad=0.16,
        fraction=0.08,
    )
    colorbar.set_label("Resident concentration (\N{MICRO SIGN}g/kg)")
    return figure


def plot_profile_curve(
    profile_simulation: pd.DataFrame,
    assessment_day: int | float,
    bottom_depth_mm: float,
) -> Figure:
    """Plot a single continuous concentration-depth profile."""

    configure_matplotlib()
    _require_columns(
        profile_simulation,
        ("time_days", "depth_mm", "concentration_ug_kg"),
        "profile_simulation",
    )
    day = float(assessment_day)
    available = np.sort(profile_simulation["time_days"].unique().astype(float))
    selected_day = available[np.argmin(np.abs(available - day))]
    selected = profile_simulation.loc[
        profile_simulation["time_days"].astype(float).eq(float(selected_day))
    ].sort_values("depth_mm")

    figure, axis = plt.subplots(figsize=(8.5, 4.5), constrained_layout=True)
    axis.plot(
        selected["concentration_ug_kg"],
        selected["depth_mm"],
        color="#006837",
        linewidth=2.2,
    )
    axis.set_ylim(float(bottom_depth_mm), 0.0)
    axis.set_xlabel("Resident concentration (\N{MICRO SIGN}g/kg)")
    axis.set_ylabel("Soil depth (mm)")
    axis.set_title(f"CLTF profile at {selected_day:g} days")
    return figure


def plot_residuals(predictions: pd.DataFrame) -> Figure:
    """Plot CLTF residuals against fitted concentration."""

    configure_matplotlib()
    return cltf_plotting.plot_residuals(predictions)


def plot_mass_fractions(
    simulation: pd.DataFrame,
    assessment_day: int | float | None = None,
) -> Figure:
    """Plot CLTF mass fractions and mark the assessment day."""

    configure_matplotlib()
    figure = cltf_plotting.plot_mass_fractions(simulation)
    return _add_assessment_to_figure(figure, assessment_day)


def plot_mass_balance(
    simulation: pd.DataFrame,
    assessment_day: int | float | None = None,
) -> Figure:
    """Plot numerical mass-balance error and mark the assessment day."""

    configure_matplotlib()
    figure = cltf_plotting.plot_mass_balance(simulation)
    return _add_assessment_to_figure(figure, assessment_day)


def plot_objective_profiles(profile: pd.DataFrame) -> Figure:
    """Plot CLTF objective profiles."""

    configure_matplotlib()
    return cltf_plotting.plot_objective_profile(profile)


def plot_bulk_density(bulk_density: pd.DataFrame) -> Figure:
    """Plot SLGA bulk-density estimates and intervals."""

    configure_matplotlib()
    return cltf_plotting.plot_bulk_density(bulk_density)


def plot_soil_properties(soil_properties: pd.DataFrame) -> Figure:
    """Plot provisional SOC and clay demo properties by depth."""

    configure_matplotlib()
    _require_columns(
        soil_properties,
        (
            "property",
            "depth_top_mm",
            "depth_bottom_mm",
            "estimate",
            "unit",
        ),
        "soil_properties",
    )
    properties = list(dict.fromkeys(soil_properties["property"].astype(str)))
    figure, axes = plt.subplots(
        1,
        len(properties),
        figsize=(8.5, 3.8),
        constrained_layout=True,
        squeeze=False,
    )
    for axis, property_name in zip(axes[0], properties):
        selected = soil_properties.loc[
            soil_properties["property"].astype(str).eq(property_name)
        ].sort_values("depth_top_mm")
        centers = (
            selected["depth_top_mm"].to_numpy(dtype=float)
            + selected["depth_bottom_mm"].to_numpy(dtype=float)
        ) / 2.0
        heights = (
            selected["depth_bottom_mm"].to_numpy(dtype=float)
            - selected["depth_top_mm"].to_numpy(dtype=float)
        )
        axis.barh(
            centers,
            selected["estimate"].to_numpy(dtype=float),
            height=heights,
            color="#56B4E9" if property_name == "SOC" else "#E69F00",
            alpha=0.78,
            edgecolor="#ffffff",
        )
        unit = selected["unit"].iloc[0]
        axis.set_title(property_name)
        axis.set_xlabel(f"{property_name} ({unit})")
        axis.set_ylim(
            soil_properties["depth_bottom_mm"].max(),
            soil_properties["depth_top_mm"].min(),
        )
        axis.set_ylabel("Soil depth (mm)")
    return figure

File: R/R/plots.R
# Script: plots.R
# Objective: Produce base-R diagnostic plots for CLTF inputs, fits, and mass balance.
# Author: Yi Yu
# Created: 2026-06-23
# Last updated: 2026-06-24
# Inputs: Climate, soil, prediction, simulation, and objective-profile tables.
# Outputs: Graphics on the active R device.
# Usage: Open a graphics device, call an exported plot function, then close the device.
# Dependencies: grDevices, graphics

cltf_plot_family <- function() {
  fonts <- names(grDevices::pdfFonts())
  if ("Arial" %in% fonts) "Arial" else "sans"
}

with_cltf_par <- function(code) {
  old <- graphics::par(no.readonly = TRUE)
  on.exit(graphics::par(old), add = TRUE)
  graphics::par(
    family = cltf_plot_family(),
    mar    = c(4.2, 4.5, 1.2, 1.0),
    las    = 1
  )
  force(code)
}

require_plot_columns <- function(data, columns, object_name) {
  missing_columns <- setdiff(columns, names(data))
  if (length(missing_columns) > 0L) {
    stop(
      object_name,
      " is missing columns: ",
      paste(missing_columns, collapse = ", "),
      call. = FALSE
    )
  }
}

cltf_layer_label <- function(depth_top_mm, depth_bottom_mm) {
  paste0(depth_top_mm, "\u2013", depth_bottom_mm, " mm")
}

#' Plot daily climate and infiltration forcing
#'
#' @param forcing Data frame containing rain, PET, daily infiltration, cumulative
#'   infiltration, and either dates or elapsed time.
#' @return The input data, invisibly.
#' @export
plot_climate_forcing <- function(forcing) {
  require_plot_columns(
    forcing,
    c(
      "rain_mm",
      "pet_mm",
      "daily_infiltration_mm",
      "cumulative_infiltration_mm"
    ),
    "forcing"
  )
  x <- if ("date" %in% names(forcing)) forcing$date else forcing$time_days
  if (is.null(x)) {
    stop("forcing must contain date or time_days.", call. = FALSE)
  }

  with_cltf_par({
    graphics::par(mfrow = c(2, 1), mar = c(3.2, 4.5, 1.2, 1.0))
    water_limit <- range(
      c(0, forcing$rain_mm, forcing$pet_mm),
      finite = TRUE
    )
    graphics::plot(
      x,
      forcing$rain_mm,
      type = "h",
      lwd  = 4,
      col  = "#0072B2",
      ylim = water_limit,
      xlab = "",
      ylab = "Daily water (mm)"
    )
    graphics::lines(x, forcing$pet_mm, col = "#D55E00", lwd = 1.8)
    graphics::legend(
      "topright",
      legend = c("Rain", "PET"),
      col    = c("#0072B2", "#D55E00"),
      lty    = 1,
      lwd    = c(4, 1.8),
      bty    = "n"
    )

    cumulative_limit <- range(
      c(0, forcing$cumulative_infiltration_mm),
      finite = TRUE
    )
    graphics::plot(
      x,
      forcing$cumulative_infiltration_mm,
      type = "l",
      lwd  = 2,
      col  = "#009E73",
      ylim = cumulative_limit,
      xlab = if (inherits(x, "Date")) "Date" else "Elapsed time (days)",
      ylab = "Cumulative infiltration (mm)"
    )
    positive <- forcing$daily_infiltration_mm > 0
    graphics::points(
      x[positive],
      forcing$cumulative_infiltration_mm[positive],
      pch = 16,
      cex = 0.65,
      col = "#009E73"
    )
  })
  invisible(forcing)
}

geometric_plot_summary <- function(predictions) {
  usable <- is.finite(predictions$analysis_concentration_ug_kg) &
    predictions$analysis_concentration_ug_kg > 0
  data <- predictions[usable, , drop = FALSE]
  key <- interaction(
    data[
      c(
        "days_since_application",
        "depth_top_mm",
        "depth_bottom_mm"
      )
    ],
    drop      = TRUE,
    lex.order = TRUE
  )
  groups <- split(data, key)
  rows <- lapply(groups, function(group) {
    data.frame(
      days_since_application = group$days_since_application[1],
      depth_top_mm           = group$depth_top_mm[1],
      depth_bottom_mm        = group$depth_bottom_mm[1],
      geometric_mean_ug_kg   = exp(mean(
        log(group$analysis_concentration_ug_kg)
      ))
    )
  })
  do.call(rbind, rows)
}

#' Plot replicate observations, geometric means, and fitted concentrations
#'
#' @param predictions Fit prediction table returned by [fit_cltf()].
#' @return The input data, invisibly.
#' @export
plot_observed_fitted <- function(predictions) {
  require_plot_columns(
    predictions,
    c(
      "days_since_application",
      "depth_top_mm",
      "depth_bottom_mm",
      "analysis_concentration_ug_kg",
      "predicted_concentration_ug_kg"
    ),
    "predictions"
  )
  positive <- c(
    predictions$analysis_concentration_ug_kg,
    predictions$predicted_concentration_ug_kg
  )
  positive <- positive[is.finite(positive) & positive > 0]
  if (length(positive) == 0L) {
    stop("predictions contain no positive concentrations to plot.", call. = FALSE)
  }
  layer <- cltf_layer_label(
    predictions$depth_top_mm,
    predictions$depth_bottom_mm
  )
  layers <- unique(layer)
  colors <- rep(c("#0072B2", "#D55E00", "#009E73", "#CC79A7"), length.out = length(layers))
  names(colors) <- layers
  geometric <- geometric_plot_summary(predictions)

  with_cltf_par({
    graphics::plot(
      range(predictions$days_since_application, finite = TRUE),
      range(positive, finite = TRUE),
      type = "n",
      log  = "y",
      xlab = "Days since application",
      ylab = "Resident concentration (\u00b5g/kg)"
    )
    for (current_layer in layers) {
      selected <- layer == current_layer
      fitted <- unique(predictions[
        selected,
        c("days_since_application", "predicted_concentration_ug_kg")
      ])
      fitted <- fitted[order(fitted$days_since_application), ]
      fitted <- fitted[
        is.finite(fitted$predicted_concentration_ug_kg) &
          fitted$predicted_concentration_ug_kg > 0,
      ]
      graphics::lines(
        fitted$days_since_application,
        fitted$predicted_concentration_ug_kg,
        col = colors[current_layer],
        lwd = 2
      )
      usable <- selected &
        is.finite(predictions$analysis_concentration_ug_kg) &
        predictions$analysis_concentration_ug_kg > 0
      graphics::points(
        predictions$days_since_application[usable],
        predictions$analysis_concentration_ug_kg[usable],
        pch = 1,
        col = colors[current_layer]
      )

      geometric_layer <- cltf_layer_label(
        geometric$depth_top_mm,
        geometric$depth_bottom_mm
      ) == current_layer
      graphics::points(
        geometric$days_since_application[geometric_layer],
        geometric$geometric_mean_ug_kg[geometric_layer],
        pch = 16,
        col = colors[current_layer]
      )
    }
    graphics::legend(
      "topright",
      legend = c(
        paste(layers, "fit"),
        "Replicate",
        "Geometric mean"
      ),
      col = c(colors[layers], "#333333", "#333333"),
      lty = c(rep(1, length(layers)), NA, NA),
      lwd = c(rep(2, length(layers)), NA, NA),
      pch = c(rep(NA, length(layers)), 1, 16),
      bty = "n",
      cex = 0.85
    )
  })
  invisible(predictions)
}

#' Plot log residuals against fitted concentration
#'
#' @param predictions Fit prediction table returned by [fit_cltf()].
#' @return The input data, invisibly.
#' @export
plot_residuals <- function(predictions) {
  require_plot_columns(
    predictions,
    c(
      "depth_top_mm",
      "depth_bottom_mm",
      "analysis_concentration_ug_kg",
      "predicted_concentration_ug_kg"
    ),
    "predictions"
  )
  residual <- if ("log_residual" %in% names(predictions)) {
    predictions$log_residual
  } else {
    log(predictions$analysis_concentration_ug_kg) -
      log(predictions$predicted_concentration_ug_kg)
  }
  usable <- is.finite(residual) &
    is.finite(predictions$predicted_concentration_ug_kg) &
    predictions$predicted_concentration_ug_kg > 0
  if (!any(usable)) {
    stop("predictions contain no finite log residuals.", call. = FALSE)
  }
  layer <- cltf_layer_label(
    predictions$depth_top_mm,
    predictions$depth_bottom_mm
  )
  layers <- unique(layer[usable])
  colors <- rep(c("#0072B2", "#D55E00", "#009E73", "#CC79A7"), length.out = length(layers))
  names(colors) <- layers

  with_cltf_par({
    graphics::plot(
      predictions$predicted_concentration_ug_kg[usable],
      residual[usable],
      log  = "x",
      pch  = 16,
      col  = colors[layer[usable]],
      xlab = "Fitted concentration (\u00b5g/kg, log scale)",
      ylab = "Log(observed) - log(fitted)"
    )
    graphics::abline(h = 0, lty = 2, col = "#666666")
    graphics::legend(
      "topright",
      legend = layers,
      col    = colors[layers],
      pch    = 16,
      bty    = "n"
    )
  })
  invisible(predictions)
}

#' Plot CLTF mass fractions through time
#'
#' @param simulation Simulation table returned by [simulate_cltf()].
#' @return The input data, invisibly.
#' @export
plot_mass_fractions <- function(simulation) {
  mass_columns <- c(
    "mass_top",
    "mass_bottom",
    "mass_below",
    "mass_degraded"
  )
  require_plot_columns(
    simulation,
    c("time_days", mass_columns),
    "simulation"
  )
  colors <- c("#0072B2", "#D55E00", "#009E73", "#777777")

  with_cltf_par({
    graphics::matplot(
      simulation$time_days,
      simulation[mass_columns],
      type = "l",
      lty  = 1,
      lwd  = 2,
      col  = colors,
      ylim = c(0, 1),
      xlab = "Days since application",
      ylab = "Applied-mass fraction"
    )
    graphics::legend(
      "right",
      legend = c("Top layer", "Bottom layer", "Below profile", "Degraded"),
      col    = colors,
      lty    = 1,
      lwd    = 2,
      bty    = "n"
    )
  })
  invisible(simulation)
}

#' Plot numerical mass balance through time
#'
#' @inheritParams plot_mass_fractions
#' @return The input data, invisibly.
#' @export
plot_mass_balance <- function(simulation) {
  mass_columns <- c(
    "mass_top",
    "mass_bottom",
    "mass_below",
    "mass_degraded"
  )
  require_plot_columns(
    simulation,
    c("time_days", mass_columns),
    "simulation"
  )
  total <- rowSums(simulation[mass_columns])
  deviation <- total - 1
  limit <- max(abs(deviation), 1e-12)

  with_cltf_par({
    graphics::plot(
      simulation$time_days,
      deviation,
      type = "l",
      lwd  = 1.8,
      col  = "#0072B2",
      ylim = c(-limit, limit),
      xlab = "Days since application",
      ylab = "Mass-balance error"
    )
    graphics::abline(h = 0, lty = 2, col = "#666666")
  })
  invisible(simulation)
}

#' Plot one or more objective profiles
#'
#' @param profile Profile table returned by [profile_cltf_parameter()].
#' @return The input data, invisibly.
#' @export
plot_objective_profile <- function(profile) {
  require_plot_columns(
    profile,
    c("parameter", "parameter_value", "objective"),
    "profile"
  )
  parameters <- unique(as.character(profile$parameter))

  with_cltf_par({
    graphics::par(
      mfrow = grDevices::n2mfrow(length(parameters)),
      mar   = c(4.2, 5.2, 1.2, 1.0)
    )
    for (parameter in parameters) {
      selected <- profile$parameter == parameter
      values <- profile[selected, ]
      values <- values[order(values$parameter_value), ]
      graphics::plot(
        values$parameter_value,
        values$objective,
        type = "b",
        pch  = 16,
        col  = "#0072B2",
        xlab = parameter,
        ylab = "Log-RMSE objective"
      )
      minimum <- which.min(values$objective)
      graphics::abline(
        v   = values$parameter_value[minimum],
        lty = 2,
        col = "#D55E00"
      )
    }
  })
  invisible(profile)
}

#' Plot SLGA bulk density by depth
#'
#' @param bulk_density Standard bulk-density band table.
#' @return The input data, invisibly.
#' @export
plot_bulk_density <- function(bulk_density) {
  require_plot_columns(
    bulk_density,
    c(
      "depth_top_mm",
      "depth_bottom_mm",
      "estimate_g_cm3",
      "lower_g_cm3",
      "upper_g_cm3"
    ),
    "bulk_density"
  )
  x_limit <- range(
    c(bulk_density$lower_g_cm3, bulk_density$upper_g_cm3),
    finite = TRUE
  )
  y_limit <- rev(range(
    c(bulk_density$depth_top_mm, bulk_density$depth_bottom_mm),
    finite = TRUE
  ))

  with_cltf_par({
    graphics::plot(
      x_limit,
      y_limit,
      type = "n",
      xlab = expression(paste("Bulk density (g/", cm^3, ")")),
      ylab = "Depth (mm)"
    )
    for (index in seq_len(nrow(bulk_density))) {
      graphics::rect(
        bulk_density$lower_g_cm3[index],
        bulk_density$depth_top_mm[index],
        bulk_density$upper_g_cm3[index],
        bulk_density$depth_bottom_mm[index],
        col    = grDevices::adjustcolor("#0072B2", alpha.f = 0.18),
        border = "#0072B2"
      )
      graphics::segments(
        bulk_density$estimate_g_cm3[index],
        bulk_density$depth_top_mm[index],
        bulk_density$estimate_g_cm3[index],
        bulk_density$depth_bottom_mm[index],
        col = "#D55E00",
        lwd = 2.5
      )
    }
    graphics::legend(
      "bottomright",
      legend = c("Estimate", "SLGA interval"),
      col    = c("#D55E00", "#0072B2"),
      lwd    = c(2.5, 8),
      bty    = "n"
    )
  })
  invisible(bulk_density)
}
